File: src/subcommands/remove.cpp
#include "version.h"
#include "../cli/color.h"
#include "../entries.h"
#include "../format.h"
#include "../parser.h"
#include "../commandline.h"
#include "subcommands.h"
#include "remove.h"

cxxopts::Options subcommands::remove::options() const
{
    auto opts = subcommands::subcommand::options();

    opts.add_options()
        ("h,help", "Prints all available options.")
        ("n", "If specified, removes the last nth entry.", cxxopts::value<int>())
        ("y,yes", "Suppresses the confirmation prompt before deletion.");

    return opts;
}

int subcommands::remove::run(const cxxopts::ParseResult& parsedOptions)
{
    if (parsedOptions.count("help") != 0) {
        return subcommands::subcommand::run(parsedOptions);
    }

    int which = 1;
    const bool suppress_confirmation = parsedOptions.count("y") != 0;

    if (parsedOptions.count("n") != 0) {
        which = parsedOptions["n"].as<int>();

        if (which <= 0) {
            std::cout << "n must be a number above 0. Use -h for help." << std::endl;

            return 1;
        }
    }

    entries::entry entry = entries::get(which);

    if (!entry.valid()) {
        std::cout << "There was no entry at position " << which << "." << std::endl;

        return 1;
    }

    if (!suppress_confirmation) {
        std::cout << "Are you sure you want to delete entry ";
        format::entry(std::cout, entry);
        std::cout << "? [y/n] ";

        if (!command_line::confirm()) {
            std::cout << "Aborting." << std::endl;
            return 0;
        }
    }

    if (entries::remove(which - 1)) {
        std::cout << "Entry removed." << std::endl;
        return 1;
    }

    std::cout << "Could not remove entry due to an unknown error." << std::endl;

    return 0;
}

std::string subcommands::remove::command() const
{
    return "remove";
}

std::string subcommands::remove::syntax() const
{
    return "[-h] [-n <integer>] [-y]";
}

std::string subcommands::remove::description() const
{
    return "Removes the last time entry or the last nth time entry.\n"
           "  Asks for confirmation before deletion.\n\n"
           "  Examples:\n"
           "    dlog remove\n"
           "    dlog remove -n 2";
}

File: src/subcommands/remove.rs
use std::{error::Error, io::Write};
use clap::Clap;
use entries::Entry;
use super::Subcommand;
use crate::{entries, errors, format};

/// Removes a time entry from the list.
///
/// EXAMPLES:
///     dlog remove       removes the last time entry
///     dlog remove 2     removes the second-to-last time entry
#[derive(Clap, Debug)]
#[clap(verbatim_doc_comment)]
pub struct Remove {
    /// Suppresses the confirmation dialog before removal.
    #[clap(short = 'y', long)]
    yes: bool,
    /// A number where the number of entries - which specifies
    /// the entry to be removed. Note that this number is
    /// one-indexed.
    #[clap(default_value = "1")]
    which: usize
}

impl Subcommand for Remove {
    fn run(&self) -> Result<(), Box<dyn Error>> {
        let mut entries = entries::read_all().unwrap_or_else(|_| Vec::<Entry>::new());

        if entries.is_empty() {
            return Err(errors::NoEntryError.into());
        }

        let len = entries.len();
        let index = entries.len() - self.which;
        let opt_entry = entries.get(index);

        if self.which >= len || opt_entry.is_none() {
            return Err(clap::Error::with_description("No entry with the specified index found.".into(), clap::ErrorKind::InvalidValue).into());
        }

        let entry = opt_entry.unwrap();
        let entry_string = if entry.complete() {
            format!("entry {}", entry)
        } else {
            format!("incomplete entry started {}", format::datetime(&entry.from))
        };

        if !self.yes {
            print!("Do you want to remove {}? (y/n) ", entry_string);
            std::io::stdout().flush()?;

            let mut input = String::new();
            std::io::stdin().read_line(&mut input)?;
            let input = input.trim().to_lowercase();
    
            if input != "y" && input != "yes" {
                println!("No entry removed.");
                return Ok(());
            }
        }

        entries.remove(index);

        entries::rewrite(&entries)?;

        println!("Removed {}", entry_string);

        Ok(())
    }
}
